contracts/SignerBondsUniswapV2.sol

```diff
@@ -0,0 +1,254 @@
+        path[0] = uniswapRouter.WETH();
+        path[1] = targetToken;
+
+        // Calculate the maximum output token amount basing on pair reserves.
+        // This value will be used as the minimum amount of output tokens that
+        // must be received for the transaction not to revert.
+        // This value includes liquidity fee equal to 0.3%.
+        uint256 amountOutMin = uniswapRouter.getAmountsOut(amount, path)[1];
+
+        require(
+            isAllowedPriceImpact(amountOutMin),
+            "Price impact exceeds allowed limit"
+        );
+
+        // Include slippage tolerance into the minimum amount of output tokens.
+        amountOutMin = amountOutMin
+            .mul(BASIS_POINTS_DIVISOR.sub(slippageTolerance))
+            .div(BASIS_POINTS_DIVISOR);
+
+        // slither-disable-next-line arbitrary-send,reentrancy-events
+        uint256[] memory amounts =
+            uniswapRouter.swapExactETHForTokens{value: amount}(
+                amountOutMin,
+                path,
+                targetTokenRecipient,
+                /* solhint-disable-next-line not-rely-on-time */
+                block.timestamp.add(swapDeadline)
+            );
+
+        emit UniswapV2SwapExecuted(amounts);
+    }
+
+    /// @notice Checks the price impact of buying a given amount of tokens
+    ///         against the maximum allowed price impact limit.
+    /// @param amount Amount of tokens.
+    /// @return True if the price impact is allowed, false otherwise.
+    function isAllowedPriceImpact(uint256 amount) internal view returns (bool) {
+        // Get reserve of the target token.
+        address WETH = uniswapRouter.WETH();
+        address token0 = WETH < targetToken ? WETH : targetToken;
+        (uint256 reserve0, uint256 reserve1, ) = uniswapPair.getReserves();
+        uint256 targetTokenReserve = WETH == token0 ? reserve1 : reserve0;
+
+        // Calculate the price impact. Multiply it by the floating point
+        // divisor to avoid float number.
+        // slither-disable-next-line divide-before-multiply
+        uint256 priceImpact =
+            CoveragePoolConstants.FLOATING_POINT_DIVISOR.mul(amount).div(
+                targetTokenReserve
+            );
+
+        // Calculate the price impact limit. Multiply it by the floating point
+        // divisor to avoid float number and make it comparable with the
+        // swap's price impact.
+        uint256 priceImpactLimit =
+            CoveragePoolConstants
+                .FLOATING_POINT_DIVISOR
+                .mul(maxAllowedPriceImpact)
+                .div(BASIS_POINTS_DIVISOR);
+
+        return priceImpact <= priceImpactLimit;
+    }
+
+    /// @notice Compute Uniswap v2 pair address.
+    /// @param factory Address of the Uniswap v2 factory.
+    /// @param tokenA Address of token A.
+    /// @param tokenB Address of token B.
+    /// @return Address of token pair.
+    function computePairAddress(
```

Comment:
Can we call [getPair](https://uniswap.org/docs/v2/smart-contracts/factory/#getpair) instead which is a view function? We would need to add `IUniswapV2Factory`, but It should save us some gas by not calculating the pair address in this contract.

Reply:
`getPair` will also consume gas because it will be called from within the contract creation transaction. It doesn't matter it's a view function. I have done some experiments using `gasleft()` and system tests. Here are the results:

The current approach:
```
computePairAddress(
    _uniswapRouter.factory(),
    _uniswapRouter.WETH(),
    targetToken
)
```
costs `7838` of gas.

The alternative:
```
IUniswapV2Factory(_uniswapRouter.factory()).getPair(_uniswapRouter.WETH(), targetToken);
```
costs `12825`.

Reply:
ok, I just looked at what gas reporter says, and indeed the cost of deployment in my implementation current vs alternative is `1605081` vs `1657021` but the function `swapSignerBondsOnUniswapV2()` diff is very little `155187` vs `155209`. Third parties like governance are interested in `swapSignerBondsOnUniswapV2()`. So I guess it boils down to code readability in this case and in large both versions are fine to me. What bugs me a tiny bit is just we copied some code from the 3rd party that can be accessed from their API. If the current approach looks fine to other devs, sure, let's leave it as is :)

Reply:
Such a change shouldn't affect `swapSignerBondsOnUniswapV2`. It uses the `uniswapPair` in the same way, no matter how it was obtained in the constructor. Regarding "copied some code from 3rd party" argument, please take a look at this doc: https://uniswap.org/docs/v2/smart-contract-integration/getting-pair-addresses. It's just one of the two ways to obtain a pair address. In this case, we're just calculating it leveraging the fact the pair contract was created using the [`CREATE2`](https://eips.ethereum.org/EIPS/eip-1014) opcode.

Reply:
Right, I'm fine with both, but the same doc as you pointed says that the "canonical" way is the first one - `getPair`.